apps/sync/js/adapters/bookmarks.js

```diff
@@ -0,0 +1,334 @@
+
+  function addBookmark(remoteRecord) {
+    // 1. Get bookmark by url(id of DataStore)
+    // 2.A. If the bookmark already exists locally,
+    //     we merge it with the remote one.
+    // 2.B Add a new record with RevisionId.
+    // 3. Add the DataStore record ID into LocalID <-> RemoteID matching table.
+
+    var id = remoteRecord.id;
+    var revisionId;
+    return _ensureStore().then(store => {
+      revisionId = store.revisionId;
+      return store.get(id);
+    }).then(localRecord => {
+      if (localRecord) {
+        var newBookmark = mergeRecordsToDataStore(localRecord, remoteRecord);
+        return store.put(newBookmark, id, revisionId);
+      }
+      return store.add(remoteRecord, id, revisionId).then(() => {
+        return setDataStoreId(remoteRecord.fxsyncId, id);
+      });
+    }).catch(e => {
+      console.error(e);
+    });
+  }
+
+  function updateBookmarks(records) {
+    return new Promise(resolve => {
+      records.reduce((reduced, current) => {
+        return reduced.then(() => {
+          if (current.deleted) {
+            return deleteBookmark(current.fxsyncId);
+          }
+          return addBookmark(current);
+        });
+      }, Promise.resolve()).then(resolve);
+    });
+  }
+
+  function deleteBookmark(fxsyncId) {
+    var url;
+    return getDataStoreId(fxsyncId).then(id => {
+      url = id;
+      return _ensureStore();
+    }).then(store => {
+      return store.remove(url);
+    });
+  }
+
+  return {
+    mergeRecordsToDataStore: mergeRecordsToDataStore,
+    setSyncedCollectionMtime: setSyncedCollectionMtime,
+    getSyncedCollectionMtime: getSyncedCollectionMtime,
+    updateBookmarks: updateBookmarks,
+    deleteBookmark: deleteBookmark
+  };
+})();
+
+DataAdapters.bookmarks = {
+/**
+    KintoCollection.list() provides a list containing all the remotely retrieved
+  Firefox Sync records sorted by "last_modified" property in descending order.
+  After each sync request we save the "last_modified" property of the last
+  processed record so we avoid going through the same records on following
+  operations.
+
+    Bookmark records are stored locally in a DataStore with format [1] while
+  bookmark records coming from Firefox Sync (via Kinto collection) have
+  format[2]. [4] is the detailed information for 6 different types defined in
+  FxSync BC. We need to convert from [1] to [2] and viceversa. Also, we need to
+  add the remote record ID[1.1] from the Kinto collection to its corresponding
+  match in the local DataStore, so we can remove the local record afterwards
+  when any deleting record requests with the format [3] are coming from FxSync.
+
+  [1] Records stored in Bookmarks DataStore (BDS): {
+    "id": "http://mozilla.org/", // KEY in PDS
+    "url": "http://mozilla.org/",
+    "name": "Mozilla",
+    "type": "url",
+    "iconable": false,
+    "icon": "http://www.lego.com/favicon.ico",
+    "fxsyncPayload": Object, // payload from BC
+    "fxsyncId": "REMOTE_ID" // [1.1]
```

Comment:
Maybe we can add `fxsyncLastModified` here and get rid of `BOOKMARKS_COLLECTION_MTIME`. Not sure which is better, though.